### DISCUSSION TITLE
Feature provided by other file: project

### DISCUSSION THREAD

**Discussion by francesco-cadei** [joaotavora/eglot#1436] [Q&A] [answered]
I can't start eglot anymore because of "Feature provided by other file: project".

**Comment by francesco-cadei** [comment]
```
Debugger entered--Lisp error: (error "Feature provided by other file: project")
  error("Feature provided by other file: %S" project)
  require-with-check(project)
  mapc(require-with-check (project flymake xref jsonrpc external-completion))
  (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion)))
  (progn (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion))))
  eval((progn (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load ... nil ...))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion)))) t)
  #f(compiled-function (&rest body) "Like `progn', but evaluates the body at compile time and at load time.\nIn interpreted code, this is entirely equivalent to `progn',\nexcept that the value of the expression may be (but is not\nnecessarily) computed at load time if eager macro expansion is\nenabled." #<bytecode 0x28f16383f4b55b6>)((let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion))))
  (eval-and-compile (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion))))
  load-with-code-conversion("/home/fra/.emacs.d/elpa/eglot-1.17/eglot.el" "/home/fra/.emacs.d/elpa/eglot-1.17/eglot.el" nil t)
  require(eglot)
  load-with-code-conversion("/home/fra/.emacs.d/elpa/flycheck-eglot-20240705.17..." "/home/fra/.emacs.d/elpa/flycheck-eglot-20240705.17..." t t)
  require(flycheck-eglot nil t)
  (not (require 'flycheck-eglot nil t))
  (if (not (require 'flycheck-eglot nil t)) (display-warning 'use-package (format "Cannot load %s" 'flycheck-eglot) :error) (condition-case err (progn (global-flycheck-eglot-mode) t) ((debug error) (funcall use-package--warning35 :config err))))
  (condition-case err (if (not (require 'flycheck-eglot nil t)) (display-warning 'use-package (format "Cannot load %s" 'flycheck-eglot) :error) (condition-case err (progn (global-flycheck-eglot-mode) t) ((debug error) (funcall use-package--warning35 :config err)))) ((debug error) (funcall use-package--warning35 :catch err)))
  (progn (use-package-ensure-elpa 'flycheck-eglot '(t) 'nil) (defvar use-package--warning35 #'(lambda (keyword err) (let ((msg (format "%s/%s: %s" ... keyword ...))) (display-warning 'use-package msg :error)))) (condition-case err (if (not (require 'flycheck-eglot nil t)) (display-warning 'use-package (format "Cannot load %s" 'flycheck-eglot) :error) (condition-case err (progn (global-flycheck-eglot-mode) t) ((debug error) (funcall use-package--warning35 :config err)))) ((debug error) (funcall use-package--warning35 :catch err))))
  (progn (progn (use-package-ensure-elpa 'flycheck-eglot '(t) 'nil) (defvar use-package--warning35 #'(lambda (keyword err) (let ((msg ...)) (display-warning 'use-package msg :error)))) (condition-case err (if (not (require 'flycheck-eglot nil t)) (display-warning 'use-package (format "Cannot load %s" 'flycheck-eglot) :error) (condition-case err (progn (global-flycheck-eglot-mode) t) ((debug error) (funcall use-package--warning35 :config err)))) ((debug error) (funcall use-package--warning35 :catch err)))))
  elisp--eval-last-sexp(nil)
  eval-last-sexp(nil)
  funcall-interactively(eval-last-sexp nil)
  command-execute(eval-last-sexp)
```

GNU Emacs 28.2 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.16.0) of 2024-06-15, modified by Debian

**Reply by joaotavora** [reply]
This is not enough, I'm sorry.  See https://joaotavora.github.io/eglot/#Troubleshooting-Eglot and focus especially on step 6. Or maybe other people will help you.

**Reply by francesco-cadei** [reply]
still unresolved
```
Compiling file /home/fra/.emacs.d/elpa/eglot-1.17/eglot.el at Sat Sep 14 13:14:59 2024
Entering directory ‘/home/fra/.emacs.d/elpa/eglot-1.17/’
eglot.el:118:1: Error: Feature provided by other file: project
```

**Reply by joaotavora** [reply]
and we're still waiting for you to provide the information we requested of you so that we can help you.  Or maybe you'll be very lucky and someone will guess. @dgutov ?

**Reply by dgutov** [reply]
I've seen this on a number of occasions when my Emacs loaded the `.el` file instead of `.elc`  on startup because the source file was newer. Then trying to load Eglot complains about this (happened with project or xref).

The way I fixed this at runtime is evaluating `(load "project.elc")` and then calling `M-x eglot` again.

**Reply by francesco-cadei** [reply]
I used `emacs -Q -f package-initialize` and opened a `.py` progect, `M-x eglot` worked fine with  `pylsp` server: i was able to do `M-?` and view `xref`.

**Reply by francesco-cadei** [reply]
`eglot` started after commenting the following lines. Do you know how to fix?
``` elisp
;; (use-package ivy-rich
;;   :config (ivy-rich-mode))
```
and
``` elisp
;; (use-package company
;;   :config
;;   (setq company-idle-delay 0
;; 	company-minimum-prefix-length 1
;; 	company-tooltip-align-annotations t)
;;   (global-company-mode))
```

**Reply by francesco-cadei** [reply, marked as answer]
solved:
``` elisp
(use-package ivy-rich)
(add-hook 'after-init-hook 'ivy-rich-mode)
```
``` elisp
(use-package company
  :config (setq company-idle-delay 0
		company-minimum-prefix-length 1
		company-tooltip-align-annotations t))
(add-hook 'after-init-hook 'global-company-mode)
```

**Reply by dgutov** [reply]
`ivy-rich` calls `(require 'project)`. Sounds like the problem depends on the package load order.

That's not great.

**Reply by johannesnauta** [reply]
I am also still having this exact same issue, but something along the lines of 
```elisp
(use-package project
  :ensure t)
```
at the top of my config still does not solve the issue. 

Specific error (which used to be a warning, but now completely bricks `eglot`), is
```
Error (use-package): eglot/:catch: Feature provided by other file: project Disable showing Disable logging
```

**Reply by joaotavora** [reply]
It's not easy to analyse any of these will precise reproducible recipes.  Also I don't know what the `use-package` does exactly.  I wish people would just type in normal use-package-less-recipes.  And show their load paths, etc. (and an `Emacs -Q` recipe is a great way to do that). All the ivy's and other packages you are using in your complex configs matter.  Anyway, I will summon @monnier here, as I believe he may have hints.  Eglot's relation with its dependencies -- including `project` -- is somewhat non-standard because of reasons.

**Reply by johannesnauta** [reply]
Yeah it for sure has to do with some packages loading other packages, and one probably loads project before it should. I am in the middle of debugging it with an MWE. If I can reproduce the bug -- which is probably after including a specific package -- I will report back.

**Reply by dgutov** [reply]
This could be due to the `require-with-check` call. I previously tracked my problem to the comparison

```
((assoc fn load-history) nil)  ;We loaded the right file.
```

where it expects the .elc file and I had the .el one loaded.

@johannesnauta Do you use the built-in package.el or something like straight/elpaca/etc, for installing packages?

**Reply by johannesnauta** [reply]
I am using `use-package` for installing and configuring packages. But I do believe it has to do with the order things are loaded. With only `package.el`, I was able to reproduce the error with `projectile`, `eglot` and `magit`. When `eglot` is loaded first, all is well, but when `magit` is loaded first, the error is raised.

So this works, but when the `magit`-part is put before `eglot`, it breaks.
```elisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)

(setq debug-on-error t)

(require 'projectile)

(unless (package-installed-p 'eglot)
  (package-refresh-contents)
  (package-install 'eglot))
(require 'eglot)

(unless (package-installed-p 'magit)
  (package-refresh-contents)
  (package-install 'magit))
(require 'magit)
```

and the error backtrace after running `emacs -Q -l init.el`
<details>
<summary>*Backtrace*</summary>

```elisp
Debugger entered--Lisp error: (error "Feature provided by other file: project")
  signal(error ("Feature provided by other file: project"))
  error("Feature provided by other file: %S" project)
  require-with-check(project)
  mapc(require-with-check (project flymake xref jsonrpc external-completion))
  (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion)))
  (progn (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion))))
  eval((progn (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load ... nil ...))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion)))) t)
  #f(compiled-function (&rest body) "Like `progn', but evaluates the body at compile time and at load time.\nIn interpreted code, this is entirely equivalent to `progn',\nexcept that the value of the expression may be (but is not\nnecessarily) computed at load time if eager macro expansion is\nenabled." #<bytecode 0x28f16383f4b6c5b>)((let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion))))
  (eval-and-compile (let ((reload (if (fboundp 'require-with-check) #'require-with-check (lambda (feature &rest _) (load (symbol-name feature) nil 'nomessage))))) (funcall reload 'eldoc nil 'reload) (funcall reload 'seq nil 'reload) (mapc reload '(project flymake xref jsonrpc external-completion))))
  eval-buffer(#<buffer  *load*-974384> nil "/home/jnauta/.emacs.d/elpa/eglot-1.17/eglot.el" nil t)  ; Reading at buffer position 6170
  load-with-code-conversion("/home/jnauta/.emacs.d/elpa/eglot-1.17/eglot.el" "/home/jnauta/.emacs.d/elpa/eglot-1.17/eglot.el" nil t)
  require(eglot)
  eval-buffer(#<buffer  *load*> nil "/home/jnauta/.emacs.d/test/init.el" nil t)  ; Reading at buffer position 437
  load-with-code-conversion("/home/jnauta/.emacs.d/test/init.el" "/home/jnauta/.emacs.d/test/init.el" nil t)
  load("/home/jnauta/.emacs.d/test/init.el" nil t)
  command-line-1(("-l" "init.el"))
  command-line()
  normal-top-level()
```
</details>

**Comment by monnier** [comment]
If the problem is `.el` vs `elc` I think we can consider this a bug in `require-with-check`, so please `M-x report-emacs-bug`.
If not, then presumably the problem is that you have two versions of `project`, installed (probably one coming from Emacs itself and another installed via something like `package.el`).  You need to make sure that the extra installation of `project` is "active" before the first time a package requires `project`, otherwise you'll end up loading the old `project` (the one from Emacs itself).

Normally this happens automatically because all the installed packages (such as `project`) get activated right before loading your init file (i.e. Emacs calls `package-activate-all` before loading `init.el`).

**Reply by johannesnauta** [reply]
There indeed seem to be two version of `project` installed, one is the built-in, and one as a dependency. But even if I load the other installation the error persists -- but tell me if I am making some mistake. 

To the snippet above, I essentially just added (before requiring any other packages)
```elisp
(add-to-list 'load-path (expand-file-name "~/.emacs.d/elpa/project-0.11.1"))
(require 'project)
```

**Reply by dgutov** [reply]
Perhaps you could double-check whether project is already loaded somehow, before these two lines even. To check, `(message "loaded? %s" (featurep 'project))`.

**Reply by johannesnauta** [reply]
It is not loaded before these two lines. `loaded? nil`

**Reply by dgutov** [reply]
Okay, to debug this further, I suggest you go to the definition of `require-with-check` (using `M-x find-function`) and add a printing statement that would print the value of `fn` (right before `cond`, for example).

Then you could search for it in `load-history` and see what the two different filename values are.

**Reply by johannesnauta** [reply]
Thanks for the help. 

After putting `(message "Requiring feature: %s from file: %s" feature fn)`, this is my `*Messages*`-buffer after doing `load-history`:

```elisp
Requiring feature: eldoc from file: /home/jnauta/.emacs.d/elpa/eldoc-1.15.0/eldoc.el
Requiring feature: seq from file: /home/jnauta/.emacs.d/elpa/seq-2.24/seq.elc
Requiring feature: project from file: /home/jnauta/.emacs.d/elpa/project-0.11.1/project.el
Eager macro-expansion failure: (error "Feature provided by other file: project")
Requiring feature: eldoc from file: /home/jnauta/.emacs.d/elpa/eldoc-1.15.0/eldoc.el
Requiring feature: seq from file: /home/jnauta/.emacs.d/elpa/seq-2.24/seq.elc
Requiring feature: project from file: /home/jnauta/.emacs.d/elpa/project-0.11.1/project.el
Eager macro-expansion failure: (error "Feature provided by other file: project")
Requiring feature: eldoc from file: /home/jnauta/.emacs.d/elpa/eldoc-1.15.0/eldoc.el
Requiring feature: seq from file: /home/jnauta/.emacs.d/elpa/seq-2.24/seq.elc
Requiring feature: project from file: /home/jnauta/.emacs.d/elpa/project-0.11.1/project.el
Entering debugger...
(("/usr/share/emacs/28.1/lisp/help-fns.elc" (require . cl-lib) (require . help-mode) (require . radix-tree) help-fns-describe-function-functions help-fns-describe-variable-functions help-fns-describe-face-functions help-fns--activated-functions help-definition-prefixes (defun . help-definition-prefixes) (defun . help--loaded-p) (defun . help--load-prefixes) ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/radix-tree.elc" (defun . radix-tree--insert) (defun . radix-tree--remove) (defun . radix-tree--lookup) (defun . radix-tree--prefixes) (defun . radix-tree--subtree) radix-tree-empty (defun . radix-tree-insert) (defun . radix-tree-lookup) (defun . radix-tree-subtree) (defun . radix-tree-prefixes) (defun . radix-tree-leaf--pcase-macroexpander) ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/cl-print.elc" cl-print-readably cl-print--number-table cl-print--currently-printing cl-print--depth (defun . cl-print-object) (cl-defmethod cl-print-object nil t t) (defun . cl-print-object-contents) (cl-defmethod cl-print-object-contents nil t t t) (cl-defmethod cl-print-object nil cons t) (cl-defmethod cl-print-object-contents nil cons t t) (cl-defmethod cl-print-object nil vector t) ...) ("/home/jnauta/.emacs.d/elpa/eldoc-1.15.0/eldoc.el" (require . cl-lib) eldoc-idle-delay eldoc-print-after-edit eldoc-echo-area-display-truncation-message eldoc-minor-mode-string eldoc-argument-case eldoc-echo-area-use-multiline-p eldoc-echo-area-prefer-doc-buffer eldoc-message-commands-table-size eldoc-message-commands eldoc-last-data ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/ert.elc" (require . cl-lib) (require . debug) (require . backtrace) (require . ewoc) (require . find-func) (require . pp) ert-batch-backtrace-right-margin (defface . ert-test-result-expected) (defface . ert-test-result-unexpected) (defun . ert-equal-including-properties) (defun . ert-test-p--cmacro) ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/pp.elc" pp-escape-newlines (t . pp-to-string) (defun . pp-to-string) (t . pp-buffer) (defun . pp-buffer) (t . pp) (defun . pp) (defun . pp-display-expression) (t . pp-eval-expression) (defun . pp-eval-expression) (t . pp-macroexpand-expression) ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/ewoc.elc" (defun . ewoc--node-left--cmacro) (defun . ewoc--node-left) (defun . ewoc--node-right--cmacro) (defun . ewoc--node-right) (defun . ewoc--node-data--cmacro) (defun . ewoc--node-data) (defun . ewoc--node-start-marker--cmacro) (defun . ewoc--node-start-marker) (defun . copy-ewoc--node) (defun . ewoc--node-create--cmacro) (defun . ewoc--node-create) ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/debug.elc" (require . cl-lib) (require . backtrace) debugger-mode-hook debugger-batch-max-lines debugger-print-function debugger-bury-or-kill debugger-step-after-exit debugger-value debugger-old-buffer debugger-previous-window debugger-previous-window-height ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/backtrace.elc" (require . find-func) (require . help-mode) (require . lisp-mode) backtrace-fontify backtrace-line-length (defun . backtrace-frame-p--cmacro) (defun . backtrace-frame-p) (defun . backtrace-frame-evald--cmacro) (defun . backtrace-frame-evald) (defun . backtrace-frame-fun--cmacro) (defun . backtrace-frame-fun) ...) ("/usr/share/emacs/28.1/lisp/emacs-lisp/find-func.elc" find-function-space-re find-function-regexp find-variable-regexp find-face-regexp find-feature-regexp find-alias-regexp (defun . find-function--defface) find-function-regexp-alist find-function-source-path find-library-source-path find-function-recenter-line ...) ("/home/jnauta/.emacs.d/elpa/magit-20240912.1705/magit.elc" (require . magit-core) (require . magit-diff) (require . magit-log) (require . magit-wip) (require . magit-apply) (require . magit-repos) (require . git-commit) (require . format-spec) (require . package) (require . with-editor) magit-openpgp-default-signing-key ...) ("/home/jnauta/.emacs.d/elpa/magit-20240912.1705/magit-autoloads.el" (t . git-rebase-current-line) (autoload . git-rebase-current-line) (t . git-rebase-mode) (autoload . git-rebase-mode) git-rebase-filename-regexp magit-define-global-key-bindings (defun . magit-maybe-define-global-key-bindings) magit-auto-revert-mode (defun . Info-follow-nearest-node@gitman) (defun . org-man-export@gitman) (t . magit-bundle) ...) ...)
```

Note that `load-history` seems to not contain `project`.

**Reply by johannesnauta** [reply]
I am perhaps also misunderstanding things, but when I omit `magit`, i.e., I only load `eglot`, the `require-with-check` function is not even called? As the above print statements do not seem to appear when only `(require 'eglot)` is provided.

**Reply by dgutov** [reply]
How are you checking that `require-with-check` is not called?

If you added a `message` statement right before `cond` (like I suggested), then it simply won't hit when `(eq lh load-history)` evaluates to non-nil.

**Reply by Xephobia** [reply]
> Okay, to debug this further, I suggest you go to the definition of `require-with-check` (using `M-x find-function`) and add a printing statement that would print the value of `fn` (right before `cond`, for example).
> 
> Then you could search for it in `load-history` and see what the two different filename values are.

I think you can know where a file is loaded by using an advice to call `debug` when `require` passed the `'project` argument. I added this at the top of my `init.el` :
```emacs-lisp
(advice-add 'require :before (lambda (feature &optional filename noerror)
			    (if (equal feature 'project)
				(debug))))
```
By pressing `c`, I see than in my case, the feature seems to be loaded by `python` (which is loaded by `ob-python` for `org-babel-load-languages`), then `projectile`, then `pet`

**Comment by monnier** [comment]
> By pressing `c`, I see than in my case, the feature seems to be loaded
>  by `python` (which is loaded by `ob-python` for
>  `org-babel-load-languages`), then `projectile`, then `pet`

The question is not just "what loads it" but also why does it load
*another* `project.el` file, which is presumably due to a change in
`load-path`.

So, you need to correlate the above occurrences of requiring `project`
with changes to `load-path`.  Does one of the above occurrences happen
before `package-activate-all` is done?

**Comment by monnier** [comment]
> By pressing `c`, I see than in my case, the feature seems to be loaded
>  by `python` (which is loaded by `ob-python` for
>  `org-babel-load-languages`), then `projectile`, then `pet`

BTW, I just pushed to Emacs's `master` a change that makes the "Feature
provided by other file" error/warning a bit more verbose: it shows both
the (new) file that would be loaded now and the file that was
previously loaded, both of which presumably provide the same feature.

**Reply by dgutov** [reply]
> BTW, I just pushed to Emacs's `master` a change that makes the "Feature
> provided by other file" error/warning a bit more verbose: it shows both
> the (new) file that would be loaded now and the file that was
> previously loaded, both of which presumably provide the same feature.

Thanks, here's the new error:

```
require-with-check: Feature ‘project’ loaded from "~/vc/emacs-master/lisp/progmodes/project.el" is now provided by "~/vc/emacs-master/lisp/progmodes/project.elc"
```

How I reproduce it:

1. Customize `load-prefer-newer` to t.
2. Edit project.el in-tree, save it, don't recompile Emacs.
3. Restart Emacs.
4. Visit some code file and try to launch `M-x eglot` there.

**Comment by Kevinizevbigie** [comment]
I have not solved this for all the packages I work with, but I got it working with the all important magit.

1. Log eglot first with use-package. `defer t`
2. then load `magit` with `defer t` and require project before it.

```lisp
(use-package eglot
  :defer t
  :config
  ;; other stuff ))

(require 'project)

(use-package magit
   :defer t
   )

(add-hook 'csharp-mode-hook 'eglot-ensure)
```

**Comment by monnier** [comment]
> ```
> require-with-check: Feature ‘project’ loaded from
> "~/vc/emacs-master/lisp/progmodes/project.el" is now provided by
> "~/vc/emacs-master/lisp/progmodes/project.elc"
> ```

Hmm... yes, this looks like a spurious case, indeed.
I guess the issue is that `locate-file` returns the `.elc` file so
`require-with-check` thinks we'd load the `.elc` file, but the `load`
would under the hood turn around and load the `.el` file instead!
So the equality check in `require-with-check` is too strict here.

**Comment by monnier** [comment]
> Hmm... yes, this looks like a spurious case, indeed.
> I guess the issue is that `locate-file` returns the `.elc` file so
> `require-with-check` thinks we'd load the `.elc` file, but the `load`
> would under the hood turn around and load the `.el` file instead!
> So the equality check in `require-with-check` is too strict here.

I opened bug#74040 for it.

**Reply by strager** [reply]
for reference: https://mail.gnu.org/archive/html/bug-gnu-emacs/2024-10/msg01544.html

**Comment by monnier** [comment]
I pushed a fix for bug@74040 to the `emacs-30` branch.
It should get merged soonish into Emacs's `master`.

**Reply by dgutov** [reply]
Thanks, Stefan!

**Reply by ed-aka-yaboi** [reply]
Has this made it into master yet?
